io.openems.edge.timeofusetariff.entsoe/src/io/openems/edge/timeofusetariff/entsoe/TouEntsoeImpl.java

```diff
@@ -0,0 +1,182 @@
+	private final AtomicReference<ImmutableSortedMap<ZonedDateTime, Float>> prices = new AtomicReference<>(
+			ImmutableSortedMap.of());
+	private final List<Consumer<Value<Integer>>> subscriptions = new ArrayList<>();
+
+	@Reference
+	private Meta meta;
+
+	private Config config = null;
+	private Currency currency;
+	private ZonedDateTime updateTimeStamp = null;
+
+	public TouEntsoeImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				TouEntsoe.ChannelId.values() //
+		);
+	}
+
+	@Activate
+	private void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+
+		if (!config.enabled()) {
+			return;
+		}
+		this.config = config;
+
+		Consumer<Currency> updateCurrency = currency -> {
+			this.currency = currency;
+			this.executor.schedule(this.task, 0, TimeUnit.SECONDS);
+		};
+
+		// Trigger to activate when the currency from Meta is updated.
+		this.subscribe(updateCurrency);
+
+		updateCurrency.accept(this.meta.getCurrencyChannel().getNextValue().asEnum());
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+		this.unsubscribeAll();
+		ThreadPoolUtils.shutdownAndAwaitTermination(this.executor, 0);
+	}
+
+	/**
+	 * Subscribes to the Currency channel to trigger on update.
+	 * 
+	 * @param updateCurrency The callback {@link Consumer}.
+	 */
+	private void subscribe(Consumer<Currency> updateCurrency) {
+
+		Consumer<Value<Integer>> subscription = currency -> {
+			updateCurrency.accept(currency.asEnum());
+		};
+		this.subscriptions.add(subscription);
+		this.meta.getCurrencyChannel().onSetNextValue(subscription);
+	}
+
+	/**
+	 * Unsubscribes from all the Subscriptions.
+	 */
+	private void unsubscribeAll() {
+		this.subscriptions.forEach(subscription -> {
+			this.meta.getCurrencyChannel().removeOnSetNextValueCallback(subscription);
+		});
+	}
+
+	private final Runnable task = () -> {
+		var token = this.config.securityToken();
+		var areaCode = this.config.biddingZone().getCode();
+		var fromDate = ZonedDateTime.now().truncatedTo(ChronoUnit.HOURS);
+		var toDate = fromDate.plusDays(1);
+		var unableToUpdatePrices = false;
+
+		try {
+			var result = EntsoeApi.query(token, areaCode, fromDate, toDate);
+
+			final double exchangeRate = this.currency == Currency.EUR //
+					? 1 // No need to fetch exchange rate from API.
+					: Utils.exchangeRateParser(ExchangeRateApi.getExchangeRates(), this.currency);
+
+			// Parse the response for the prices
+			this.prices.set(Utils.parse(result, "PT60M", exchangeRate));
+
+			// store the time stamp
+			this.updateTimeStamp = ZonedDateTime.now();
+		} catch (IOException | ParserConfigurationException | SAXException | OpenemsNamedException e) {
+			e.printStackTrace();
+			if (e instanceof OpenemsNamedException) {
+				this.logWarn(this.log, "Unable to get the currency exchange rate " + e.getMessage());
```

Comment:
No, distinguishing the error message by the Exception type is a hack. Don't to this. You'd better create a proper Exception message before and just log it here (`this.logWarn(this.log, e.getMessage());`). 

It's good practice to log the `e.printStackTrace();` after a descriptive warning and not before.